File: src/WhisperHeim/Services/Templates/FuzzyMatcher.cs
namespace WhisperHeim.Services.Templates;

/// <summary>
/// Provides fuzzy string matching for matching spoken template names
/// against stored template names. Uses Levenshtein distance normalized
/// by the longer string length.
/// </summary>
public static class FuzzyMatcher
{
    /// <summary>
    /// Finds the best matching template name from the candidates.
    /// Returns null if no candidate scores above the threshold.
    /// </summary>
    /// <param name="spoken">The spoken/transcribed text to match.</param>
    /// <param name="candidates">Available template names.</param>
    /// <param name="threshold">Minimum similarity score (0.0 to 1.0). Default 0.4.</param>
    /// <returns>The best matching candidate name, or null if none matched.</returns>
    public static string? FindBestMatch(string spoken, IEnumerable<string> candidates, double threshold = 0.4)
    {
        if (string.IsNullOrWhiteSpace(spoken))
            return null;

        var normalizedSpoken = Normalize(spoken);
        string? bestMatch = null;
        double bestScore = 0;

        foreach (var candidate in candidates)
        {
            if (string.IsNullOrWhiteSpace(candidate))
                continue;

            var normalizedCandidate = Normalize(candidate);

            // Try exact containment first (highest priority)
            if (normalizedSpoken.Contains(normalizedCandidate, StringComparison.OrdinalIgnoreCase) ||
                normalizedCandidate.Contains(normalizedSpoken, StringComparison.OrdinalIgnoreCase))
            {
                double containScore = 0.95;
                if (containScore > bestScore)
                {
                    bestScore = containScore;
                    bestMatch = candidate;
                }
                continue;
            }

            // Compute similarity via Levenshtein distance
            var score = ComputeSimilarity(normalizedSpoken, normalizedCandidate);
            if (score > bestScore)
            {
                bestScore = score;
                bestMatch = candidate;
            }
        }

        return bestScore >= threshold ? bestMatch : null;
    }

    /// <summary>
    /// Computes a similarity score between 0.0 and 1.0 for two strings.
    /// 1.0 = identical, 0.0 = completely different.
    /// </summary>
    public static double ComputeSimilarity(string a, string b)
    {
        if (string.Equals(a, b, StringComparison.OrdinalIgnoreCase))
            return 1.0;

        var maxLen = Math.Max(a.Length, b.Length);
        if (maxLen == 0)
            return 1.0;

        var distance = LevenshteinDistance(a.ToLowerInvariant(), b.ToLowerInvariant());
        return 1.0 - (double)distance / maxLen;
    }

    private static string Normalize(string input)
    {
        // Strip everything except letters and digits, then lowercase.
        // This makes matching ignore case, spaces, punctuation, and hyphens.
        var chars = input.Where(char.IsLetterOrDigit).ToArray();
        return new string(chars).ToLowerInvariant();
    }

    /// <summary>
    /// Computes the Levenshtein edit distance between two strings.
    /// </summary>
    private static int LevenshteinDistance(string s, string t)
    {
        var n = s.Length;
        var m = t.Length;

        if (n == 0) return m;
        if (m == 0) return n;

        // Use single-row optimization
        var previousRow = new int[m + 1];
        var currentRow = new int[m + 1];

        for (var j = 0; j <= m; j++)
            previousRow[j] = j;

        for (var i = 1; i <= n; i++)
        {
            currentRow[0] = i;

            for (var j = 1; j <= m; j++)
            {
                var cost = s[i - 1] == t[j - 1] ? 0 : 1;
                currentRow[j] = Math.Min(
                    Math.Min(currentRow[j - 1] + 1, previousRow[j] + 1),
                    previousRow[j - 1] + cost);
            }

            // Swap rows
            (previousRow, currentRow) = (currentRow, previousRow);
        }

        return previousRow[m];
    }
}

File: mac_version/whisperheim/services/templates/fuzzy_matcher.py
"""Fuzzy string matching for matching spoken template names against stored names.

Ported from Windows WhisperHeim FuzzyMatcher.cs.
Uses Levenshtein distance normalized by the longer string length.
"""

import re
from typing import Optional


class FuzzyMatcher:
    """Provides fuzzy string matching for voice-triggered template lookup."""

    @staticmethod
    def find_best_match(
        spoken: str,
        candidates: list[str],
        threshold: float = 0.4,
    ) -> Optional[str]:
        """Find the best matching template name from candidates.

        Returns None if no candidate scores above the threshold.

        Args:
            spoken: The spoken/transcribed text to match.
            candidates: Available template names.
            threshold: Minimum similarity score (0.0 to 1.0). Default 0.4.
        """
        if not spoken or not spoken.strip():
            return None

        normalized_spoken = FuzzyMatcher._normalize(spoken)
        best_match: Optional[str] = None
        best_score: float = 0.0

        for candidate in candidates:
            if not candidate or not candidate.strip():
                continue

            normalized_candidate = FuzzyMatcher._normalize(candidate)

            # Try exact containment first (highest priority)
            if (normalized_candidate in normalized_spoken
                    or normalized_spoken in normalized_candidate):
                contain_score = 0.95
                if contain_score > best_score:
                    best_score = contain_score
                    best_match = candidate
                continue

            # Compute similarity via Levenshtein distance
            score = FuzzyMatcher.compute_similarity(
                normalized_spoken, normalized_candidate
            )
            if score > best_score:
                best_score = score
                best_match = candidate

        return best_match if best_score >= threshold else None

    @staticmethod
    def compute_similarity(a: str, b: str) -> float:
        """Compute a similarity score between 0.0 and 1.0 for two strings.

        1.0 = identical, 0.0 = completely different.
        """
        if a.lower() == b.lower():
            return 1.0

        max_len = max(len(a), len(b))
        if max_len == 0:
            return 1.0

        distance = FuzzyMatcher._levenshtein_distance(a.lower(), b.lower())
        return 1.0 - distance / max_len

    @staticmethod
    def _normalize(text: str) -> str:
        """Strip everything except letters and digits, then lowercase."""
        return re.sub(r"[^a-zA-Z0-9]", "", text).lower()

    @staticmethod
    def _levenshtein_distance(s: str, t: str) -> int:
        """Compute the Levenshtein edit distance between two strings."""
        n = len(s)
        m = len(t)

        if n == 0:
            return m
        if m == 0:
            return n

        # Single-row optimization
        previous_row = list(range(m + 1))
        current_row = [0] * (m + 1)

        for i in range(1, n + 1):
            current_row[0] = i
            for j in range(1, m + 1):
                cost = 0 if s[i - 1] == t[j - 1] else 1
                current_row[j] = min(
                    current_row[j - 1] + 1,      # insertion
                    previous_row[j] + 1,          # deletion
                    previous_row[j - 1] + cost,   # substitution
                )
            previous_row, current_row = current_row, previous_row

        return previous_row[m]
